include/osquery/error.h

```diff
@@ -0,0 +1,140 @@
+/**
+ *  Copyright (c) 2018-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under both the Apache 2.0 license (found in the
+ *  LICENSE file in the root directory of this source tree) and the GPLv2 (found
+ *  in the COPYING file in the root directory of this source tree).
+ *  You may select, at your option, one of the above-listed licenses.
+ */
+
+#pragma once
+
+#include <boost/core/demangle.hpp>
+#include <memory>
+#include <new>
+#include <string>
+#include <typeinfo>
+
+namespace osquery {
+class Status;
+
+class ErrorBase {
+ public:
+  virtual std::string getShortMessage() const = 0;
+  virtual std::string getFullMessage() const = 0;
+  virtual std::string getShortMessageRecursive() const = 0;
+  virtual std::string getFullMessageRecursive() const = 0;
+  virtual ~ErrorBase(){};
+};
+
+template <class T>
+class Error : public ErrorBase {
+ private:
+  static std::string getErrorTypeName() {
+    return boost::core::demangle(typeid(T).name());
+  }
+
+ public:
+  explicit Error(T error_code,
+                 std::string message = "",
+                 std::shared_ptr<ErrorBase> underlying_error = nullptr)
+      : errorCode_(error_code),
+        message_(std::move(message)),
+        underlyingError_(std::move(underlying_error)) {}
+
+  Error(Error&& other) {
+    errorCode_ = other.errorCode_;
+    message_ = std::move(other.message_);
+    underlyingError_ = std::move(other.underlyingError_);
+  }
+
+  Error& operator=(Error&& other) {
+    if (this != &other) {
+      errorCode_ = other.errorCode_;
+      message_ = std::move(other.message_);
+      underlyingError_ = std::move(other.underlyingError_);
+    }
+    return *this;
+  }
+
+  T getErrorCode() const {
+    return errorCode_;
+  }
+
+  virtual std::shared_ptr<ErrorBase> getUnderlyingError() const {
+    return underlyingError_;
+  }
+
+  virtual std::string getShortMessage() const {
+    return getErrorTypeName() + " " +
+           std::to_string(static_cast<int>(errorCode_));
+  }
+
+  virtual std::string getFullMessage() const {
+    std::string full_message = getShortMessage();
+    if (message_.size() > 0) {
+      full_message += " (" + message_ + ")";
```

Comment:
nit: For performance please append individually, like:

```
full_message += " (";
full_message += message_;
full_message += ")";
```

Reply:
Full message is slow path, so IMHO performance is not important here

Reply:
You could use `boost::format` here to avoid both performance penalty and verbosity.

Reply:
I think boost::format will be slower than current implemnetaion, becuase boost::format is not performance oriented and slower than iostream